jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/PerMessageDeflateExtension.java

```diff
@@ -464,16 +464,18 @@ private boolean inflate(Frame frame, Callback callback, boolean first) throws Da
             chunk.setPayload(payload);
             chunk.setFin(frame.isFin() && complete);
 
-            boolean succeedCallback = complete;
+            boolean completeCallback = complete;
             AtomicReference<ByteBuffer> payloadRef = _payloadRef;
             Callback payloadCallback = Callback.from(() ->
             {
                 getBufferPool().release(payloadRef.getAndSet(null));
-                if (succeedCallback)
+                if (completeCallback)
                     callback.succeeded();
             }, t ->
             {
                 getBufferPool().release(payloadRef.getAndSet(null));
+                if (completeCallback)
+                    callback.failed(t);
                 failFlusher(t);
             });
```

Comment:
I'm not exactly following the logic here, which means that it probably needs a comment.
Specifically in the case of complete==true, it is not clear who actually does complete the callback passed into the method.

I also see the atomic payloadRef, is that just protecting from calls to both succeeded and failed?  Or is it racing with something else?
If it is just checking that the other method has not been called, then the other actions should be conditional on a non-null return from the getAndSet, so you don't call the nested callback twice either nor fail the flusher.

Reply:
If `completeCallback` is true it nulls out the frame callback field after this method so the `payloadCallback` must always complete the frame callback.

If `completeCallback` is false it will not have been cleared out so the `failFlusher()` call will clear the callback when it reaches `onCompleteFailure()`.

The atomic payloadRef is for a different race, can't remember the exact details but will be described in the review comments here #7360.

Reply:
probably best to say this in comments rather than here.

The comments in #7360 indicate that the atomic is for the prevention of multiple calls, so other actions should be conditional on a atomic as well, else we do some things correctly once, but others twice.  Since we have the atomic, we should use it!

Reply:
Note that the `Callback` is an instance of `CountingCallback` and can also be called more than once and uses an `AtomicInteger`.

Reply:
@lachlan-roberts I think it could be something like:
```java
            Callback payloadCallback = Callback.from(() ->
            {
                ByteBuffer buffer = payloadRef.getAndSet(null);
                if (buffer == null)
                    return; 
                getBufferPool().release(buffer);
                if (completeCallback)
                    callback.succeeded();
            }, t ->
            {
                ByteBuffer buffer = payloadRef.getAndSet(null);
                if (buffer == null)
                    return; 
                getBufferPool().release(buffer);
                if (completeCallback)
                    callback.failed(t);
                failFlusher(t);
            });

```

Reply:
> Note that the `Callback` is an instance of `CountingCallback` and can also be called more than once and uses an `AtomicInteger`.

Yeah but surely it is only called once from each wrapping?

Reply:
Either way... this callback needs some javadoc to explain why the atomics are used and who completes the nested callback if completeCallback is false.

Reply:
This code suggestion will not work. We cannot do the early return if `buffer == null` otherwise the callback will never be completed.

I have now added a lengthy comment above this section of code.